_posts/2023-08-25-using-kani-to-validate-security-boundaries-in-aws-firecracker.md

```diff
@@ -0,0 +1,292 @@
+
+For the past 10 months, Felipe Monteiro, an Applied Scientist on the [Kani](https://github.com/model-checking/kani/) team and Patrick Roy, a Software Development Engineer from the [AWS Firecracker](https://github.com/firecracker-microvm/firecracker) team, collaborated to develop Kani harnesses that verify safety-critical properties in the Firecracker Virtual Machine Monitor. Firecracker is an open source project written in Rust which uses the Linux Kernel-based Virtual Machine (KVM) to create and manage microVMs. Firecracker has a minimalist design which allows fast (~150ms) microVM start-up time, secure multi-tenancy of microVMs on the same host and memory/CPU over-subscription. Firecracker is currently used in production by AWS Lambda, AWS Fargate and parts of AWS Analytics to build their service platforms. 
+
+As a result of this collaboration, the Firecracker team is now running multiple verification suites in their continuous integration pipelines, ensuring that all checked properties of critical systems are upheld on every code change.
+
+In this blog post, we show how Kani helped Firecracker harden two core components, namely our I/O rate limiter and transport layer (VirtIO), presenting the issues we were able to identify and fix. Particularly, the second part of this post picks up from a [previous Kani/Firecracker blogpost](https://model-checking.github.io/kani-verifier-blog/2022/07/13/using-the-kani-rust-verifier-on-a-firecracker-example.html) and shows how improvements to Kani over the last year made verifying conformance with a section of the VirtIO specification feasible.
+
+## Noisy-Neighbor Mitigations via Rate Limiting
+
+In multi-tenant systems, microVMs from different customers simultaneously co-exist on the same physical host. We thus need to ensure that access to host resources, such as disk and network, is shared fairly. We should not allow a single “greedy” microVM to transfer excessive amounts of data from disk to the point where other microVMs’ disk access gets starved off (a “noisy neighbor” scenario). Firecracker offers a mitigation for this via *I/O rate-limiting*. From the [documentation](https://github.com/firecracker-microvm/firecracker/blob/4a3e9bd76d9fc57a3538c1aeb7e5687de43a0efa/docs/design.md#io-storage-networking-and-rate-limiting):
+
+
+>Firecracker provides VirtIO/block and VirtIO/net emulated devices, along with the application of rate limiters to each volume and network interface to make sure host hardware resources are used fairly by multiple microVMs. These are implemented using a token bucket algorithm [...].
+
+
+In a token bucket based rate-limiter, each microVM has a budget of “tokens” that can be exchanged for permission to do one byte of I/O. These tokens regenerate at a fixed rate, and if the microVM runs out of tokens, it gets I/O-throttled.
```

Comment:
In this explanation, you never refer to the image. Would be helpful to use it in the explanation at some point? It'd feel more integrated with the text.